language/tools/move-unit-test/src/test_reporter.rs

```diff
@@ -0,0 +1,245 @@
+    }
+
+    pub fn aborted(abort_code: u64) -> Self {
+        FailureReason::Aborted("Test was not expected to abort".to_string(), abort_code)
+    }
+
+    pub fn timeout() -> Self {
+        FailureReason::Timeout("Test timed out".to_string())
+    }
+
+    pub fn unknown() -> Self {
+        FailureReason::Unknown("ITE: An unknown error was reported.".to_string())
+    }
+}
+
+impl TestFailure {
+    pub fn new(failure_reason: FailureReason, fname: &str, vm_error: Option<VMError>) -> Self {
+        Self {
+            failure_reason,
+            function_ident: fname.to_owned(),
+            vm_error,
+            storage_state: None,
+        }
+    }
+
+    pub fn render_error(&self, test_plan: &TestPlan) -> String {
+        match &self.failure_reason {
+            FailureReason::NoAbort(message) => message.to_string(),
+            FailureReason::Timeout(message) => message.to_string(),
+            FailureReason::WrongAbort(message, expected_code, other_code) => {
+                let base_message = format!(
+                    "{}. Expected test to abort with {} but instead it aborted with {} here",
+                    message, expected_code, other_code,
+                );
+                Self::report_abort(test_plan, base_message, &self.vm_error)
+            }
+            FailureReason::Aborted(message, code) => {
+                let base_message = format!("{} but it aborted with {} here", message, code);
+                Self::report_abort(test_plan, base_message, &self.vm_error)
+            }
+            FailureReason::Unknown(message) => {
+                format!(
+                    "{}. VMError (if there is one) is: {}",
+                    message,
+                    self.vm_error
+                        .as_ref()
+                        .map(|err| format!("{:#?}", err))
+                        .unwrap_or_else(|| "".to_string())
+                )
+            }
+        }
+    }
+
+    fn report_abort(
+        test_plan: &TestPlan,
+        base_message: String,
+        vm_error: &Option<VMError>,
+    ) -> String {
+        let report_error = if control::SHOULD_COLORIZE.should_colorize() {
+            errors::report_errors_to_color_buffer
+        } else {
+            errors::report_errors_to_buffer
+        };
+
+        let vm_error = match vm_error {
+            None => return base_message,
+            Some(vm_error) => vm_error,
+        };
+
+        match vm_error.location() {
+            Location::Module(module_id) => {
+                let errors = vm_error
+                    .offsets()
+                    .iter()
+                    .filter_map(|(fdef_idx, offset)| {
+                        let function_source_map = test_plan
+                            .module_info
+                            .get(&module_id)
+                            .and_then(|info| info.1.get_function_source_map(*fdef_idx).ok());
+                        function_source_map.and_then(|function_source_map| {
```

Comment:
Can't these just be `?`. I feel like that might be a bit mroe readable than a series of `and_then` + `map`

Reply:
The issue is that these are inside of a `filter_map` so we don't want to just return but instead maintain the "`None`-ness" of them throughout this process. 

Let me take a closer look to see if I can alter the flow of things here to make this easier.

Reply:
Maintaining "`None`-ness" is precisely what `?` would do.

Reply:
Woah...TIL. I didn't realize that this would work inside of a `filter_map` like that! All the times I've used `?` inside of an iterator it will cause the `collect` type to be a `Result` or `Option` and I didn't realize it would work in a different way here...

Reply:
You were likely doing a normal `map`. Anyway in that case you can specify collect to do `collect::<Vec<_>>` which will likely get you `Vec<Option<_>>` instead of `Option<Vec<_>>` or vice versa if you want the other